[Refactoring] Relax MUX select signal checks for out-of-order Aligner support

DanaKossaybati opened the pull request:
In the original backend design, **MUX** components are assumed to be placed only at the **top of a basic block (BB)**. Under this assumption, the **type verification system** enforces that the **inputs**, **output**, and **select** signal of a MUX must all have **matching extra signals** (e.g., tags).

However, **out-of-order execution** introduces a more complex use of MUXes as part of an `Aligner` operation. These MUXes do **not necessarily appear at the top of a BB**, and their select signals originate from a different source: the **tag output of an `Untagger`**.

Here is what a typical `Aligner` looks like:

<img src="https://github.com/user-attachments/assets/05bf4a47-9dd3-4e3e-9f8b-da1c0cbad1d2" width="500" />

In the case where **extra signals (tags)** are involved, the `Aligner` takes this form:

<img src="https://github.com/user-attachments/assets/7285240f-9374-48da-a13b-c6affe2d6548" width="500" />

Consider the following:

- When **extra signals** (like tags) are present(e.g., due to data flowing through an outer **tagged region**)they are typically attached to the **data** paths.
- The **`Untagger`** operation removes the tag from the data, but **does not** pass those extra signals to its **tag output**.

As a result, the select signal from the `Untagger`'s tag output **lacks the extra tag signals**, while the inputs and output of the MUX still **retain** them. This creates a mismatch if the original type-checking rule is enforced.

### Why this matters:

If the type system continues to require **select = inputs = output** in terms of extra signals, the MUX inside the Aligner would be **invalid** under the original rules—even though it's functionally correct in this out-of-order context.

### Therefore:

The **MUXes used in out-of-order `Aligner` operations** must **relax** the original constraint:  
> The **select** signal should **not** be required to carry the same extra signals as the **inputs** and **output**.

This allows the system to accept valid MUXes where the select is untagged, and the data lines are tagged—correctly reflecting the semantics of the `Untagger`.

murphe67 commented:
This situation does not look any different to any other MUX to me (it would be strange if it did, since that would imply this is not actually a MUX.)

You have simply gone for our original solution for speculative MUXes here, and for the other solution with other MUXes (we switched to the second solution for spec MUXes, because they both work).

Either MUX conditions should not have extra signals, or they should have the same as the data inputs.

![mux_condition_types](https://github.com/user-attachments/assets/bb9fd3a3-a3a7-4283-9ea7-f35a382965ea)


The same applies to your demux:
![demux_condition_types](https://github.com/user-attachments/assets/dd838b1d-f341-4a2f-9676-c705fbac59b7)

Whatever we do for one, we should do for all.

AyaElAkhras commented:
I agree that Demuxes can share the same requirements. However, as of now,

1. we're the only ones currently using Demuxes, so we've tailored the type verification and backend to suit our needs,
2. we have only one use case for Demuxes — as part of the Aligner implementation — where they _always have no extra signals in their condition_, regardless of the number of extra signals in the input and output data.

Let's first agree on what needs to be done for Muxes. Once that's clear, we can adjust our Demuxes if necessary. I also suggest we hold off on discussing CMerges until we've settled on the best approach for Muxes.

For Muxes, we have the following requirements:

**Requirement 1:** Assuming we do not use Muxes to implement an Aligner (e.g., we implement it in pure RTL instead of composing elastic units), then a tagged Mux has identical extra signals on all of its inputs (including the condition input) and its output. How shall we implement such a Mux?

**Requirement 2:** We want to implement an Aligner using elastic components. More precisely, we use Untaggers, Muxes and Demuxes to implement it, as shown in Dana's figures above. Before presenting the requirement, let's clarify the operation of an Untagger: 

Every Untagger takes a single input and has two outputs, as shown here. Its single input should have at least one extra signal. One output is identical to the input but with one less extra signal and the second output has that one extra signal as a part of its data field (so data without extra signal). 

![untagger(1)](https://github.com/user-attachments/assets/f9e1adf7-3c67-4af6-a0a8-69f479228f38)

Back to the requirement, part of implementing an Aligner is by adding a Mux that has its condition driven from the second output of the Untagger (the one that has one tag in the data field with no extra signals at all). If we have no hierarchical tagging, we have no problem because this Mux will have no extra signals at its inputs (both data and condition inputs) and outputs, as shown here.

![no-hierarchical](https://github.com/user-attachments/assets/4ef75cef-4783-405a-a3bf-6556c55e761e)

However, if we have hierarchical tagging, this Mux will have extra signals at its data inputs and outputs and no extra signals at its condition input, as shown here. How shall we implement such a Mux?

![yes-hierarchical](https://github.com/user-attachments/assets/1d6c8fe7-7272-4de7-ab45-936d1313370b)

Do you have an implementation that already satisfies these 2 requirements?

murphe67 commented:
Apologies for not engaging with how you have framed the discussion, but I don't think it would be too productive  😅

There are two possible solutions to your problem: 1) type verification enforces that the mux condition input has the same extra signals as the data input, or 2) type verification enforces that the mux condition has no extra signals.

Either option is fine, but I directly oppose "sometimes" having extra signals on a channel. It makes verification more complex, it makes generating the netlist more complex, it makes generating the operation RTL more complex.

In this situation, it is also unnecessary: we have two good solutions that you can choose between.

I can definitely meet Monday if you'd like- the diagrams I drew above do cover both options fully also, maybe they help to organize any questions you have?

DanaKossaybati commented:
> Let's meet on Monday then. Is 2 PM good for you? Is it good for @shundroid and @DanaKossaybati?

Yes, 2pm works for me!👍

shundroid commented:
Also sorry for the late reply, I also prefer 4pm today if it suits you.

AyaElAkhras commented:
No worries, yup 4 pm works. See you later! :)

DanaKossaybati commented:
I’ve updated the type verification system in our branch ([out-of-order-execution](https://github.com/EPFL-LAP/dynamatic/tree/out-of-order-execution)) with the following changes:
-     The select signal of a `MUX` should no longer carry extra signals.
-     The index output of a `CMERGE` should also not include any extra signals.

To support this, I introduced a new operation, `ExtractOp`, which removes extra signals from a channel. This operation should be used at the select of a `MUX` if extra signals are present, in order to avoid violating the updated type verification rules.

These changes are now part of the [Out-of-Order Execution PR](https://github.com/EPFL-LAP/dynamatic/pull/357).

Please make any necessary adjustments to the speculation pass or other parts of main that might conflict with the updated type verification system.
With this, I’ll go ahead and close this PR.
